### ISSUE TITLE
Microprose F117 - bad frequency

### ISSUE THREAD

**Issue by andrich48** [IllidanS4/pcsconv#3]
Problems with Microprose F-117 Stealth Fighter game, intro movie.
OS - Windows 7 x64, Modified DOSBox, pcsconv v1.3.
After capturing intro movie and running pcsconv with parameters:

> D:\Games\dosbox\pcsconv_13>pcsconv.exe -in pcs -out play f117.pcs.txt test.wav
> pcsconv 1.3 (c) 2013 - 2016 by IllidanS4@gmail.com
> test.wav
> 00:01:53.7820000
> Creating WAV...
> Playing...

Resulting music is very weird and different from real music (from in-game movie).
When running pcsconv with 'beep' parameter, lots of errors are written in console:

> D:\Games\dosbox\pcsconv_13>pcsconv.exe -in pcs -out beep f117.pcs.txt test.wav
> pcsconv 1.3 (c) 2013 - 2016 by IllidanS4@gmail.com
> test.wav
> 00:01:53.7820000
> Bad frequency 35
> Bad frequency 20
> Bad frequency 20
> Bad frequency 25
> Bad frequency 25
> Bad frequency 25
> Bad frequency 25
> Bad frequency 25
> ^C:00:09.2907713

Conversion to rpc goes without errors, but result is same - weird wav sound, errors with beeps.
It seems, errors are generated in places like here:

> SetCountdown: 58712
> Delay: 2
> SetCountdown: 58713
> Delay: 3

I can attach my version of F117, if needed.
[f117.pcs.txt](https://github.com/IllidanS4/pcsconv/files/532046/f117.pcs.txt)

**Comment by IS4Code** [comment]
The "Bad frequency" message is reported if the frequency is outside [37; 32767] range for Console.Beep. If so, it is simply ignored and has no effect on the result. The range corresponds to countdown of [37; 32248]. I have added some code that removed these frequencies, but I don't have the game and I don't know how it should sound. Please test the latest version with the fixes here:
[pcsconv.zip](https://github.com/IllidanS4/pcsconv/files/532144/pcsconv.zip)
Use "--sanitize" to remove bad frequencies, and tell me if it sounds better or more accurate. If not, please also include the sound that DOSBox produces in the game, for me to compare.

**Comment by andrich48** [comment]
Sorry i couldnt add that argument :( Where to add it? I will try more later.

> D:\Games\dosbox\pcsconv_14>pcsconv.exe --input-type pcs --output-type play --san
> itize f117.pcs.txt test.wav
> pcsconv 1.4 (c) 2013 - 2016 by IllidanS4@gmail.com
> Invalid path. Try --help.
> 
> D:\Games\dosbox\pcsconv_14>pcsconv.exe --sanitize -in pcs -out play f117.pcs.txt
> test.wav
> pcsconv 1.4 (c) 2013 - 2016 by IllidanS4@gmail.com
> Unexpected switch test.wav. Try --help.

My DOSBox output is attached to 1st message in this thread, file named 'f117.pcs.txt'.

**Comment by IS4Code** [comment]
I am sorry, I have added the argument wrong way. This this:
[pcsconv.zip](https://github.com/IllidanS4/pcsconv/files/533248/pcsconv.zip)
The correct way to run it is:
pcsconv.exe --input-type pcs --output-type play --sanitize f117.pcs.txt
or
pcsconv.exe --input-type pcs --output-type play f117.pcs.txt --sanitize
In the first version I sent you, only the second method worked. Output file is not needed at all if you use the "play" output type, all wave rendering happens in memory then. Order of all parameters can also be changed freely.

**Comment by andrich48** [comment]
No changes with new key, music still weird :(
Original wav gzipped soundtrack attached here.

[f117_intro.wav.gz](https://github.com/IllidanS4/pcsconv/files/537271/f117_intro.wav.gz)

**Comment by IS4Code** [comment]
Now I see that for this whole time, the routines that handle correct PC Speaker mechanics were flawed. Where the DOS program set the frequency, I turned prematurely the PC Speaker on, while I should have waited till it actually turns the PC Speaker on using its specific port. This error was in the ProcessPCS method. Now it stores the frequency and uses it just when the PC Speaker is turned on (if it isn't already), and not before it.

[pcsconv.zip](https://github.com/IllidanS4/pcsconv/files/537554/pcsconv.zip)

This produces accurate pauses matching the sound you provided.

**Comment by andrich48** [comment]
Thanks! It seems to be working, I hear no differences between original wav and re-rendered.
Are all latest changes merged? With all those '--sanitize', '--click' parameters.
Also maybe its time to update documentation, '-h' does not print all possible keys.

**Comment by IS4Code** [comment]
Yes, all changes are contained in this last executable. Official release will come soon. I also plan to update the documentation.

**Comment by andrich48** [comment]
Sorry I forgot, '--sanitize' removes wrong frequency beep for '-out beep' mode only or for '-out wav' mode too? I think impossible sounds should be rendered in WAV if '--sanitize' parameter not set, just for compatibility, because original MS-DOS somehow manages to make low frequency sounds.
Also, for my personal pc-speaker, frequency '37' sounds perfect and technically my low-cost speaker can generate all low unsupported tones.

**Comment by IS4Code** [comment]
Yes, "--sanitize" removes these sounds from all outputs, because it is an input filter. So even if you use "-out txt" and "--sanitize", it will also remove them. The range 37 - 32767 was specified by Microsoft for some reason (and the reason may not in fact be related to audibility at all), so I decided to abide by it.

I have done some little research, and the reasons behind the limits are more of a technical than physiological. 32767 is the upper limit of a signed 16-bit integer, and 37 corresponds to a highest signed 16-bit countdown value (~1193180/32767). Though the numbers are not arbitrary, they still appear to be arbitrarily chosen limits, because technically a wider range of frequencies can be supported (19 to 1193180 Hz). You are right that this range is not the "hearing range", so either the limits should be changed, or the message.

But the limits for "--sanitize" were chosen so because you complained about the "Bad frequency" message for the beep output type. Lowering the minimum value would make it ineffective, so the only solution is to either change the message text, or remove it completely. For the moment, I have modified the "hearing range" to "standard range", though making it's a bit ambiguous.
